# `rule-name`

Please describe the origin of the rule here.

## Rule Details

This rule aims to...

:x: Examples of **incorrect** code for this rule:

```js
// fill me in
```

:white_check_mark: Examples of **correct** code for this rule:

```js
// fill me in
```

## Options

If this rule does not have options, this section can be removed. Show all the valid options in the `type Options` code, along with their default values below. Give examples for each option in a h3.

This rule has an object option.

```ts
type Options = {
  /**
   * This in an option that can be set and this is what it does/means.
   */
  anOption: string,
};

const defaults = {
  anOption: 'some-value'
};
```

### `anOption`

:x: Examples of **incorrect** code for this rule with the `{ "anOption": "some-value" }` option:

```html
<!-- fill me in -->
```

:white_check_mark: Examples of **correct** code for this rule with the `{ "anOption": "some-value" }` option:

```html
<!-- fill me in -->
```

## When Not To Use It

Give a short description of when it would be appropriate to turn off this rule.

## Further Reading

If there are other links that describe the issue this rule addresses, please include them here in a bulleted list.
